langstream-core/src/main/java/ai/langstream/impl/parser/ModelBuilder.java

```diff
@@ -126,18 +149,62 @@ public static ApplicationWithPackageInfo buildApplicationInstance(
                                         e);
                             }
                         }
+                    } else {
+                        if (Files.isDirectory(path)) {
+                            String filename = path.getFileName().toString();
+                            if (filename.equals("python")) {
+                                recursiveAppendFiles(sortedPyBinaries, path, path);
+                            } else {
+                                if (filename.equals("java")) {
+                                    final Path lib = path.resolve("lib");
+                                    if (Files.exists(lib) && Files.isDirectory(lib)) {
+                                        recursiveAppendFiles(sortedJavaBinaries, lib, lib);
+                                    }
+                                }
+                            }
+                        }
                     }
                 }
             }
         }
+        final ApplicationWithPackageInfo applicationWithPackageInfo =
+                buildApplicationInstance(applicationContents, instanceContent, secretsContent);
+        if (!sortedJavaBinaries.isEmpty()) {
+            final byte[] data = binariesDigestMapper.writeValueAsBytes(sortedJavaBinaries);
+            applicationWithPackageInfo.javaBinariesDigest = checksumFunction.apply(data);
+        }
+        if (!sortedPyBinaries.isEmpty()) {
+            final byte[] data = binariesDigestMapper.writeValueAsBytes(sortedPyBinaries);
+            applicationWithPackageInfo.pyBinariesDigest = checksumFunction.apply(data);
+        }
+        return applicationWithPackageInfo;
+    }
+
+    private static void recursiveAppendFiles(
+            TreeMap<String, String> sortedPyBinaries, Path current, Path rootPath)
+            throws IOException {
 
-        return buildApplicationInstance(applicationContents, instanceContent, secretsContent);
+        if (Files.isRegularFile(current)) {
+            try {
+                final Path relativeCurrent = rootPath.relativize(current);
+                sortedPyBinaries.put(
+                        relativeCurrent.toString(),
+                        Files.readString(current, StandardCharsets.UTF_8));
```

Comment:
this doesn't work for binary files, like zip (jar) files
you can keep the byte array here